docs/src/lucet-runtime/killswitch.md

```diff
@@ -0,0 +1,346 @@
+
+In this chapter we will describe both a typical usage of `KillSwitch` as a
+mechanism to enforce execution time limits, and the implementation complexities
+underlying `KillSwitch` correctness.
+
+`KillSwitch` are valid for termination only on the instance call after they are
+created, and until an instance is reset. When a call into a guest returns, the
+shared state by which `KillSwitch` signal is replaced, and an attempt to
+`terminate` will fail with an `Err`.
+
+## Example
+
+This example taken from `lucet_runtime_tests::timeout::timeout_in_guest`:
+```rust
+let module = mock_timeout_module();
+let region = TestRegion::create(1, &Limits::default()).expect("region can be created");
+let mut inst = region
+    .new_instance(module)
+    .expect("instance can be created");
+let kill_switch = inst.kill_switch();
+
+// Spawn a thread to terminate the instance after waiting for 100ms.
+let t = thread::Builder::new()
+    .name("killswitch".to_owned())
+    .spawn(move || {
+        thread::sleep(Duration::from_millis(100));
+        assert_eq!(kill_switch.terminate(), Ok(KillSuccess::Signalled));
+    })
+    .expect("can spawn a thread");
+
+// Begin running the instance, which will be terminated remotely by the KillSwitch.
+match inst.run("infinite_loop", &[]) {
+    Err(Error::RuntimeTerminated(TerminationDetails::Remote)) => {
+        // the result of a guest that was remotely terminated (a KillSwitch at work)
+    }
+    res => panic!("unexpected result: {:?}", res),
+}
+
+// wait for the KillSwitch-firing thread to complete
+t.join().unwrap();
+```
+
+## Implementation
+
+As this section discusses implementation details of `lucet-runtime`, it
+will refer to structures and enums that are unexported.  Since there are
+no `lucet-runtime` paths for these items, this section will occasionally
+use fully-qualified paths to the `lucet_runtime_internals` path where.
+they are originally defined.
+
+As much as is possible, `KillSwitch` tries to be self-contained; no members are
+public, and it tries to avoid leaking details of its state-keeping into public
+interfaces. Currently, `lucet-runtime` is heavily dependent on POSIX
+thread-directed signals to implement guest timeouts - for non-POSIX platforms
+alternate implementations may be plausible, but need careful consideration of
+the race conditions that can arise from other platform-specific functionality.
+
+`KillSwitch` fundamentally relies on two piece of state for safe operation,
+which are encapsulated in a `KillState` held by the `lucet_runtime::Instance`
+it terminates:
+* `execution_domain`, a `Domain` that describes the kind of execution that is
+  currently happening in the `Instance`. This is kept in a `Mutex` since in
+  many cases it will need to be accessed either by a `KillSwitch` or
+  `KillState`, and for the duration either are considering the domain, it must
+  block other users.
+* `terminable`, an `AtomicBool` that indicates if the `Instance` may stop
+  executing.
```

Comment:
This description might benefit from mentioning how we are using `terminable` as a flag, for signal safety reasons.